src/FlexVotingClient.sol

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity >=0.8.10;
+
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {Checkpoints} from "@openzeppelin/contracts/utils/Checkpoints.sol";
+import {IFractionalGovernor} from "src/interfaces/IFractionalGovernor.sol";
+import {IVotingToken} from "src/interfaces/IVotingToken.sol";
+
+// TODO natspec
+// this is a contract to make it easy to build clients for flex voting governors.
+abstract contract FlexVotingClient {
+  using SafeCast for uint256;
+  using Checkpoints for Checkpoints.History;
+
+  /// @notice The voting options corresponding to those used in the Governor.
+  enum VoteType {
+    Against,
+    For,
+    Abstain
+  }
+
+  /// @notice Data structure to store vote preferences expressed by depositors.
+  struct ProposalVote {
+    uint128 againstVotes;
+    uint128 forVotes;
+    uint128 abstainVotes;
+  }
+
+  /// @notice Map proposalId to an address to whether they have voted on this proposal.
+  mapping(uint256 => mapping(address => bool)) private proposalVotersHasVoted;
+
+  /// @notice Map proposalId to vote totals expressed on this proposal.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @notice The governor contract associated with this governance token. It
+  /// must be one that supports fractional voting, e.g. GovernorCountingFractional.
+  IFractionalGovernor public immutable GOVERNOR;
+
+  /// @notice Mapping from address to stored (not rebased) balance checkpoint history.
+  mapping(address => Checkpoints.History) private balanceCheckpoints;
+
+  /// @notice History of total stored (not rebased) balances.
+  Checkpoints.History internal totalDepositCheckpoints;
+
+  /// @dev Constructor.
+  /// @param _governor The address of the flex-voting-compatible governance contract.
+  constructor(address _governor) {
+    GOVERNOR = IFractionalGovernor(_governor);
+  }
+
+  /// @notice Returns the _user's current balance in storage. If the balance
+  /// rebases this should return the non-rebased value, i.e. the value before
+  /// any computations are run or interest is applied.
+  function _rawBalanceOf(address _user) internal view virtual returns (uint256);
+
+  // TODO add natspec
+  function selfDelegate() public {
```

Comment:
should this be `internal` because it's called at initialization/construction? not clear why it should be public

Reply:
Agree, good catch. Done in 82c2ba0